universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java

```diff
@@ -2,55 +2,98 @@
 
 import static com.google.common.base.Preconditions.checkNotNull;
 
+import com.jayway.jsonpath.Configuration;
 import com.jayway.jsonpath.DocumentContext;
 import com.jayway.jsonpath.JsonPath;
-import com.jayway.jsonpath.PathNotFoundException;
+import com.jayway.jsonpath.Option;
+import com.jayway.jsonpath.spi.mapper.MappingException;
 import java.time.Instant;
+import java.util.HashMap;
 import java.util.Locale;
 import java.util.Objects;
 import java.util.Optional;
 import javax.annotation.Nullable;
 
 public class ApplicantData {
+  // Suppress errors thrown by JsonPath and instead return null if a path does not exist in a JSON
+  // blob.
+  private static final Configuration CONFIGURATION =
+      Configuration.defaultConfiguration().addOptions(Option.SUPPRESS_EXCEPTIONS);
 
   private DocumentContext jsonData;
   private static final String EMPTY_APPLICANT_DATA_JSON = "{ \"applicant\": {}, \"metadata\": {} }";
 
   public ApplicantData() {
-    this(JsonPath.parse(EMPTY_APPLICANT_DATA_JSON));
+    this(EMPTY_APPLICANT_DATA_JSON);
   }
 
-  public ApplicantData(DocumentContext jsonData) {
-    this.jsonData = checkNotNull(jsonData);
+  public ApplicantData(String jsonData) {
+    this.jsonData = JsonPath.using(CONFIGURATION).parse(checkNotNull(jsonData));
   }
 
   public Locale preferredLocale() {
     return Locale.ENGLISH;
   }
 
   /**
-   * Attempts to read a string at the given path in the applicant's answer data. Returns an {@code
-   * Optional.empty()} if the path does not exist.
+   * Puts the given value at the given path in the underlying JSON data. Builds up the necessary
+   * structure along the way, i.e., creates parent objects where necessary.
    *
-   * <p>The design suggestion for these methods is to have one per {@link
-   * services.question.ScalarType}
+   * @param path the {@link Path} with the fully specified path, e.g., "applicant.favorites.color"
+   *     or the equivalent "$.applicant.favorite.colors".
+   * @param value the value to place; values of type Map will create the equivalent JSON structure
    */
-  public Optional<String> readString(String path) {
+  public void put(Path path, Object value) {
```

Comment:
Suggestion (blocking): please make this method private and instead expose public methods with more specific typing for the value to be written than `Object` e.g. `writeString(Path path, String string)` and `writeDate(Path path, Date date)` . This will help prevent bugs where the wrong type could be written to a given path.

Reply:
Thanks for calling out why this matters! I added methods for String, Integer, and Objects (aka Maps) for now

I would love to use `Path` in Questions too - would that be an appropriate refactor?

Reply:
I think it would, but maybe it can be done as a follow-on PR? Up to you.